### ISSUE TITLE
pylint crashes with astroid.exceptions._NonDeducibleTypeHierarchy

### ISSUE THREAD

**Issue by mckelvin** [PyCQA/pylint#1739]
### Steps to reproduce
1. Prepare `reproduce.py` as follows
```python
from unknown_module import UnknownClass


class KnownClass(object):
    pass


class ExtendedUnknownClass(UnknownClass):
    pass


variable = KnownClass()
variable += ExtendedUnknownClass()

variable.what_ever_attribute_called
```
2. Run `pylint reproduce.py`

### Current behavior

The `pylint` will crash with the following traceback:

```python
Traceback (most recent call last):
  File "/usr/local/bin/pylint", line 11, in <module>
    load_entry_point('pylint==1.8.0', 'console_scripts', 'pylint')()
  File "/usr/local/lib/python2.7/site-packages/pylint/__init__.py", line 13, in run_pylint
    Run(sys.argv[1:])
  File "/usr/local/lib/python2.7/site-packages/pylint/lint.py", line 1326, in __init__
    linter.check(args)
  File "/usr/local/lib/python2.7/site-packages/pylint/lint.py", line 747, in check
    self._do_check(files_or_modules)
  File "/usr/local/lib/python2.7/site-packages/pylint/lint.py", line 880, in _do_check
    self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
  File "/usr/local/lib/python2.7/site-packages/pylint/lint.py", line 959, in check_astroid_module
    walker.walk(ast_node)
  File "/usr/local/lib/python2.7/site-packages/pylint/utils.py", line 994, in walk
    self.walk(child)
  File "/usr/local/lib/python2.7/site-packages/pylint/utils.py", line 994, in walk
    self.walk(child)
  File "/usr/local/lib/python2.7/site-packages/pylint/utils.py", line 991, in walk
    cb(astroid)
  File "/usr/local/lib/python2.7/site-packages/pylint/checkers/typecheck.py", line 676, in visit_attribute
    inferred = list(node.expr.infer())
  File "/usr/local/lib/python2.7/site-packages/astroid/decorators.py", line 85, in wrapped
    res = next(generator)
  File "/usr/local/lib/python2.7/site-packages/astroid/bases.py", line 95, in _infer_stmts
    for inferred in stmt.infer(context=context):
  File "/usr/local/lib/python2.7/site-packages/astroid/context.py", line 40, in cache_generator
    for result in generator:
  File "/usr/local/lib/python2.7/site-packages/astroid/decorators.py", line 85, in wrapped
    res = next(generator)
  File "/usr/local/lib/python2.7/site-packages/astroid/context.py", line 40, in cache_generator
    for result in generator:
  File "/usr/local/lib/python2.7/site-packages/astroid/decorators.py", line 85, in wrapped
    res = next(generator)
  File "/usr/local/lib/python2.7/site-packages/astroid/inference.py", line 409, in _filter_operation_errors
    for result in infer_callable(self, context):
  File "/usr/local/lib/python2.7/site-packages/astroid/inference.py", line 732, in _infer_augassign
    for result in results:
  File "/usr/local/lib/python2.7/site-packages/astroid/inference.py", line 625, in _infer_binary_operation
    context, reverse_context)
  File "/usr/local/lib/python2.7/site-packages/astroid/inference.py", line 600, in _get_aug_flow
    elif helpers.is_subtype(left_type, right_type):
  File "/usr/local/lib/python2.7/site-packages/astroid/helpers.py", line 145, in is_subtype
    return _type_check(type2, type1)
  File "/usr/local/lib/python2.7/site-packages/astroid/helpers.py", line 132, in _type_check
    raise exceptions._NonDeducibleTypeHierarchy
astroid.exceptions._NonDeducibleTypeHierarchy
```

### Expected behavior

It should not crash.

### pylint --version output

I tried both the version in master(6ef5895) by `pip install git+https://github.com/PyCQA/pylint.git` and the 1.7.4 on PyPI.

```
$ pylint --version
No config file found, using default configuration
pylint 1.8.0,
astroid 1.5.3
Python 2.7.13 (default, Jul 18 2017, 09:17:00)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]

$ pylint --version
No config file found, using default configuration
pylint 1.7.4,
astroid 1.5.3
Python 2.7.13 (default, Jul 18 2017, 09:17:00)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

Both of them can reproduce the issue.

**Comment by mckelvin** [comment]
I believe this issue has been resolved in https://github.com/PyCQA/astroid/pull/443 (verified using `git bisect`). But it's not released yet.

**Comment by PCManticore** [comment]
This is indeed solved by PyCQA/astroid#443. We're going to release it this month as pylint 1.8 and astroid 1.6

**Comment by mckelvin** [comment]
Great to hear that. Thanks.
